File: frontend/src/types/market.ts
/** Common market data types for stock_dash */

export interface StockQuote {
  symbol: string;
  name: string;
  price: number;
  change: number;
  changePercent: number;
  volume: number;
  high: number;
  low: number;
  open: number;
  prevClose: number;
  timestamp: string;
  market: "KR" | "US";
}

export interface CryptoQuote {
  symbol: string;
  name: string;
  price: number;
  change: number;
  changePercent: number;
  volume24h: number;
  high24h: number;
  low24h: number;
  timestamp: string;
  currency: "KRW" | "USD";
}

export interface PreciousMetalQuote {
  metal: "gold" | "silver";
  pricePerOz: number;
  pricePerGram: number;
  change: number;
  changePercent: number;
  currency: "USD" | "KRW";
  timestamp: string;
}

export interface ExchangeRate {
  baseCurrency: string;
  quoteCurrency: string;
  rate: number;
  change: number;
  changePercent: number;
  timestamp: string;
}

export interface MarketIndex {
  symbol: string;
  name: string;
  value: number;
  change: number;
  changePercent: number;
  market: "KR" | "US" | "JP" | "EU";
  isOpen: boolean;
  timestamp: string;
}

export interface NewsArticle {
  id: string;
  title: string;
  summary: string;
  source: string;
  url: string;
  imageUrl?: string;
  publishedAt: string;
  relatedSymbols: string[];
  category: string;
}

export interface StockDetail {
  symbol: string;
  name: string;
  price: number;
  change: number;
  changePercent: number;
  volume: number;
  high: number;
  low: number;
  open: number;
  prevClose: number;
  week52High: number;
  week52Low: number;
  marketCap?: number;
  market: "KR" | "US";
  timestamp: string;
}

export interface OHLCData {
  time: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume?: number;
}

export type WidgetType =
  | "stock-chart"
  | "crypto-price"
  | "precious-metal"
  | "market-index"
  | "exchange-rate"
  | "news-feed"
  | "stock-detail";

export interface DashboardWidget {
  id: string;
  type: WidgetType;
  title: string;
  config: Record<string, unknown>;
  position: { x: number; y: number; w: number; h: number };
}

File: backend/app/schemas/market.py
"""Market data schemas for API responses."""

from datetime import datetime, timezone
from typing import Generic, TypeVar, Optional
from pydantic import BaseModel, field_validator

T = TypeVar("T")


class ApiResponse(BaseModel, Generic[T]):
    """Standard API response wrapper."""
    data: T
    status: str = "success"
    message: Optional[str] = None
    timestamp: str = ""

    @field_validator("timestamp", mode="before")
    @classmethod
    def set_default_timestamp(cls, v: str) -> str:
        return v or datetime.now(timezone.utc).isoformat()


class StockQuote(BaseModel):
    symbol: str
    name: str
    price: float
    change: float
    changePercent: float
    volume: int
    high: float
    low: float
    open: float
    prevClose: float
    timestamp: str
    market: str  # "KR" | "US"


class CryptoQuote(BaseModel):
    symbol: str
    name: str
    price: float
    change: float
    changePercent: float
    volume24h: float
    high24h: float
    low24h: float
    timestamp: str
    currency: str = "KRW"


class PreciousMetalQuote(BaseModel):
    metal: str  # "gold" | "silver"
    pricePerOz: float
    pricePerGram: float
    change: float
    changePercent: float
    currency: str = "USD"
    timestamp: str


class ExchangeRate(BaseModel):
    baseCurrency: str
    quoteCurrency: str
    rate: float
    change: float
    changePercent: float
    timestamp: str


class MarketIndex(BaseModel):
    symbol: str
    name: str
    value: float
    change: float
    changePercent: float
    market: str  # "KR" | "US"
    isOpen: bool
    timestamp: str


class NewsArticle(BaseModel):
    id: str
    title: str
    summary: str
    source: str
    url: str
    imageUrl: Optional[str] = None
    publishedAt: str
    relatedSymbols: list[str] = []
    category: str = "general"


class StockDetail(BaseModel):
    symbol: str
    name: str
    price: float
    change: float
    changePercent: float
    volume: int
    high: float
    low: float
    open: float
    prevClose: float
    week52High: float
    week52Low: float
    marketCap: Optional[float] = None
    market: str  # "KR" | "US"
    timestamp: str


class OHLCData(BaseModel):
    time: str
    open: float
    high: float
    low: float
    close: float
    volume: Optional[float] = None
